File: backend/app/models/coupon.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Float
from app.db.base import Base, TimestampMixin

class Coupon(TimestampMixin, Base):
    __tablename__ = "coupons"

    id = Column(Integer, primary_key=True, autoincrement=True)
    code = Column(String(50), unique=True, index=True, nullable=False)
    discount_percent = Column(Float, nullable=False, default=10.0) # Percentage e.g. 10.0 for 10%
    is_active = Column(Boolean, nullable=False, default=True)
    valid_until = Column(DateTime, nullable=True)
    max_uses = Column(Integer, nullable=False, default=1)
    uses_count = Column(Integer, nullable=False, default=0)

File: frontend/src/types/coupon.ts
export interface CouponValidationResponse {
    valid: boolean;
    discount_percent: number;
    code: string;
}

export interface CouponCreatePayload {
    code: string;
    discount_percent: number;
    valid_until?: string | null;
    max_uses: number;
}

export interface CouponCreateResponse {
    message: string;
    coupon: {
        code: string;
        discount: number;
    };
}
